Libraries/oneMKL/monte_carlo_european_opt/src/montecarlo_main.cpp

```diff
@@ -0,0 +1,157 @@
+        sycl::usm_allocator<DataType, sycl::usm::alloc::shared> alloc(my_queue);
+        std::vector<DataType, decltype(alloc)> h_CallResult(num_options, alloc);
+        std::vector<DataType, decltype(alloc)> h_CallConfidence(num_options, alloc);
+        std::vector<DataType, decltype(alloc)> h_StockPrice(num_options, alloc);
+        std::vector<DataType, decltype(alloc)> h_OptionStrike(num_options, alloc);
+        std::vector<DataType, decltype(alloc)> h_OptionYears(num_options, alloc);
+        DataType* h_CallResult_ptr = h_CallResult.data();
+        DataType* h_CallConfidence_ptr = h_CallConfidence.data();
+        DataType* h_StockPrice_ptr = h_StockPrice.data();
+        DataType* h_OptionStrike_ptr = h_OptionStrike.data();
+        DataType* h_OptionYears_ptr = h_OptionYears.data();
+
+        // calculate the number of blocks
+        constexpr DataType fpath_lengthN = static_cast<DataType>(path_length);
+        constexpr DataType stddev_denom = 1.0 / (fpath_lengthN * (fpath_lengthN - 1.0));
+        DataType confidence_denom = 1.96 / std::sqrt(fpath_lengthN);
+
+        constexpr int rand_seed = 777;
+        constexpr std::size_t local_size = 256;
+        const std::size_t global_size = (num_options * local_size) / ITEMS_PER_WORK_ITEM; // It requires num_options be divisible by ITEMS_PER_WORK_ITEM
+        const int block_n = path_length / (local_size * VEC_SIZE);
+
+        timer<timer_enabled> tt{};
+        double total_time = 0.0;
+
+        namespace mkl_rng = oneapi::mkl::rng;
+
+        mkl_rng::philox4x32x10 engine(my_queue, rand_seed); // random number generator object
+        auto rng_event_1 = mkl_rng::generate(mkl_rng::uniform<DataType>(5.0, 50.0), engine, num_options, h_StockPrice_ptr);
+        auto rng_event_2 = mkl_rng::generate(mkl_rng::uniform<DataType>(10.0, 25.0), engine, num_options, h_OptionStrike_ptr);
+        auto rng_event_3 = mkl_rng::generate(mkl_rng::uniform<DataType>(1.0, 5.0), engine, num_options, h_OptionYears_ptr);
+
+        std::size_t n_states = global_size;
+        using EngineType =
+#if USE_MCG59
+            mkl_rng::device::mcg59<VEC_SIZE>;
+#elif USE_MRG
+            mkl_rng::device::mrg32k3a<VEC_SIZE>;
+#else
+            mkl_rng::device::philox4x32x10<VEC_SIZE>;
+#endif
+
+        // initialization needs only on first step
+        auto deleter = [my_queue](auto* ptr) {sycl::free(ptr, my_queue);};
+        auto rng_states_uptr = std::unique_ptr<EngineType, decltype(deleter)>(sycl::malloc_device<EngineType>(n_states, my_queue), deleter);
+        auto* rng_states = rng_states_uptr.get();
+
+        my_queue.parallel_for(
+            sycl::range<1>(n_states),
+            std::vector<sycl::event>{rng_event_1, rng_event_2, rng_event_3},
+            [=](sycl::item<1> idx) {
+                auto id = idx[0];
+#if USE_MRG
+                constexpr std::uint32_t seed = 12345u;
+                rng_states[id] = EngineType({ seed, seed, seed, seed, seed, seed }, { 0, (4096 * id) });

```

Comment:
Do we use exponential skip only for MRG on purpose? It can work as line 89 too, but init will be slower